File: elixir/connect/lib/connect.ex
defmodule Connect do
  @doc """
  Calculates the winner (if any) of a b
  using "O" as the white player
  and "X" as the black player
  """
  @spec result_for([String.t()]) :: :none | :black | :white
  def result_for(b) do
    cond do
      0..(length(b) - 1) |> Enum.any?(fn r ->
        path?(b, r, 0, ?X, [], &reached_black_goal/3) end) -> :black
      0..(width(b) - 1)  |> Enum.any?(fn c ->
        path?(b, 0, c, ?O, [], &reached_white_goal/3) end) -> :white
      true -> :none
    end
  end

  defp width(b), do: String.length(Enum.at(b, 0))
  defp on_board?(b, r, c), do: r >= 0 && r < length(b) && c >= 0 && c < width(b)
  defp entry(b, r, c), do: b |> Enum.at(r) |> to_charlist |> Enum.at(c)
  defp path?(b, r, c, stone, visited, reached_player_goal) do
    cond do
      visited |> Enum.member?({r, c})                    -> false
      not(on_board?(b, r, c)) or entry(b, r, c) != stone -> false
      reached_player_goal.(b, r, c)                      -> true
      true                                               ->
        [{-1, 0},{-1, 1},{0, -1},{0, 1},{1, -1},{1, 0}] |> Enum.any?(fn {dr, dc} ->
          path?(b, r + dr, c + dc, stone, [{r, c}|visited], reached_player_goal)
        end)
    end
  end
  defp reached_black_goal(b, _r, c), do: c == width(b) - 1
  defp reached_white_goal(b, r, _c), do: r == length(b) - 1
end

File: fsharp/connect/Connect.fs
module Connect

type Player = White | Black 

type private Board = string list 

let private offsets = [| (-1, -1); (1, -1); (0, -2); (0, 2); (-1, 1); (1, 1) |]

let winner (board: Board) : Player option =
    let height = board |> List.length
    let width = ((board[0] |> String.length) - (height - 1)) / 2 + 1
    let onBoard r c =  r >= 0 && r <= height - 1 && c >= r && c <= r + 2 * (width - 1)
    let rec isConnected r c stone visited toEndEdge =
        if visited |> List.contains (r, c) then false
        else 
            if not (onBoard r c) || board[r][c] <> stone then false
            elif toEndEdge (r, c) then true 
            else offsets |> Seq.exists (fun (dr, dc) ->
                isConnected (r + dr) (c + dc) stone ((r, c) :: visited) toEndEdge)
    let toRightEdge (r, c) = c = r + 2 * (width - 1)
    let toBottomEdge (r, _ ) = r = height - 1 
    if  [0..(height - 1)] |> Seq.exists (fun r -> isConnected r r 'X' [] toRightEdge) then
        Some Black
    elif [0..(width - 1)] |> Seq.exists (fun c -> isConnected 0 (2 * c) 'O' [] toBottomEdge) then 
        Some White 
    else None